### DISCUSSION TITLE
Project: ophyd - bluesky-compatible device abstractions for EPICS

### DISCUSSION THREAD

**Discussion by klauer** [pcdshub/epics-codeathon-2022-python#9] [Projects]
Source code: https://github.com/bluesky/ophyd
Documentation: https://blueskyproject.io/ophyd/

> Ophyd is Python library for interfacing with hardware. It provides an abstraction layer than enables experiment orchestration and data acquisition code to operate above the specifics of particular devices and control systems.
> 
> Ophyd is typically used with the [Bluesky Run Engine](http://blueskyproject.io/bluesky) for experiment orchestration and data acquistion. It is also sometimes used in a stand-alone fashion.
> 
> Many facilities use ophyd to integrate with control systems that use [EPICS](http://www.aps.anl.gov/epics/) , but ophyd's design and some of its objects are also used to integrate with other control systems.
> 
>  *  Put the details specific to a device or control system behind a high-level interface with methods like trigger(), read(), and set(...).
>  *  Group individual control channels (such as EPICS V3 PVs) into logical "Devices" to be configured and used as units with internal coordiantion.
> *   Assign readings with names meaningful for data analysis that will propagate into metadata.
> *  Categorize readings by "kind" (primary reading, configuration, engineering/debugging) which can be read selectively.

Related: 
* https://github.com/pcdshub/epics-codeathon-2022-python/discussions/8

**Comment by klauer** [comment]
## Project idea

* Provide introduction to ophyd for codeathon members
* Collaboratively sift through numerous issues: https://github.com/bluesky/ophyd/issues/
* Reproduce and propose a solution to the issue
* Nudge reviewers or help finish up existing PRs: https://github.com/bluesky/ophyd/pulls

**Comment by ZLLentz** [comment]
There are a number of features we've added to pcdsdevices that are good candidates for ophyd: https://github.com/pcdshub/pcdsdevices/issues/796

**Reply by ZLLentz** [reply]
I'd like to take a stab at upstreaming the simplest/cleanest utilities:
- https://github.com/pcdshub/pcdsdevices/blob/82de60f361814ede14e021d8ca03ce4e98f14b9d/pcdsdevices/pv_positioner.py#L9
- https://github.com/pcdshub/pcdsdevices/blob/82de60f361814ede14e021d8ca03ce4e98f14b9d/pcdsdevices/pv_positioner.py#L71
- https://github.com/pcdshub/pcdsdevices/blob/82de60f361814ede14e021d8ca03ce4e98f14b9d/pcdsdevices/pv_positioner.py#L98
- https://github.com/pcdshub/pcdsdevices/blob/82de60f361814ede14e021d8ca03ce4e98f14b9d/pcdsdevices/signal.py#L764

**Reply by ZLLentz** [reply]
A bit harder, requires some extra thinking: https://github.com/bluesky/ophyd/issues/971

**Reply by ZLLentz** [reply]
First PR for this topic, for `InternalSignal`: https://github.com/bluesky/ophyd/pull/1045

**Comment by ZLLentz** [comment]
I think this is another simple addition that would be useful: `EpicsSignalWO` https://github.com/bluesky/ophyd/issues/1021

**Reply by tangkong** [reply]
The existing conversations touch on some EPICS details that I don't fully understand, meaning I'd be interested in seeing this one worked on.

**Reply by tangkong** [reply]
I'm taking a stab at this.  Will throw up a draft PR once it's in a more complete state

**Reply by tangkong** [reply]
Draft PR for `EpicsSignalWO`: https://github.com/bluesky/ophyd/pull/1044

**Comment by ZLLentz** [comment]
This documentation issue from ophyd is also a good candidate for a codeathon project: https://github.com/bluesky/ophyd/issues/995

**Reply by tangkong** [reply]
Maybe we could also loop back on the flyer documentation? https://github.com/bluesky/ophyd/issues/1000

**Reply by tangkong** [reply]
I think Thomas Morris (?) was going to give this a look

**Reply by mrakitin** [reply]
Ping @thomaswmorris

**Comment by ZLLentz** [comment]
Ophyd log levels for the device-level logs are very noisy (self.log)- lots of "INFO" logs that should probably be "DEBUG". Perhaps we could take a look at that during this week? I might make an issue for it later.

**Reply by klauer** [reply]
👍 for addressing this so we can remove our logger helper

**Reply by ZLLentz** [reply]
I'm going to tackle this right now

**Reply by ZLLentz** [reply]
https://github.com/bluesky/ophyd/pull/1047

**Comment by mrakitin** [comment]
https://github.com/bluesky/ophyd/pull/961 is ready for another round of review. I addressed the questions raised previously by the reviewers. This can be useful for the identification of the places where the specific classes are coming from, specifically, when defined in IPython startup files.
